File: linux-client/osm_ride_linux/route/brouter_client.py
"""Routes a sequence of tapped waypoints onto real roads/paths using BRouter's free public
routing API (https://brouter.de) - same technique as
app/src/main/java/com/ewaldmire/osmride/route/BRouterClient.kt.

Synchronous (stdlib urllib, no extra dependency) - callers on an asyncio event loop (the GTK UI)
should wrap calls with `await asyncio.to_thread(route, waypoints)` to avoid blocking.
"""

from __future__ import annotations

import urllib.error
import urllib.parse
import urllib.request

from .models import RouteWaypoint

_BASE_URL = "https://brouter.de/brouter"
_PROFILE = "trekking"
_CONNECT_TIMEOUT_SECONDS = 30


class BRouterError(Exception):
    pass


def route(waypoints: list[RouteWaypoint]) -> str:
    """Returns the raw routed GPX text following roads through `waypoints`, in order."""
    if len(waypoints) < 2:
        raise BRouterError("Need at least 2 waypoints to route")

    lonlats = "|".join(f"{wp.lon},{wp.lat}" for wp in waypoints)
    query = urllib.parse.urlencode(
        {"lonlats": lonlats, "profile": _PROFILE, "alternativeidx": 0, "format": "gpx"}
    )
    url = f"{_BASE_URL}?{query}"

    try:
        with urllib.request.urlopen(url, timeout=_CONNECT_TIMEOUT_SECONDS) as response:
            return response.read().decode("utf-8")
    except urllib.error.HTTPError as e:
        error_body = e.read().decode("utf-8", errors="replace")
        raise BRouterError(f"BRouter request failed ({e.code}): {error_body}") from e
    except urllib.error.URLError as e:
        raise BRouterError(f"BRouter request failed: {e.reason}") from e

File: app/src/main/java/com/ewaldmire/osmride/route/BRouterClient.kt
package com.ewaldmire.osmride.route

import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable

/** A tapped point placed by the user while building a route in-app. */
@Serializable
data class RouteWaypoint(val lat: Double, val lon: Double)

/**
 * Routes a sequence of tapped waypoints onto real roads/paths using BRouter's free public
 * routing API (https://brouter.de) - the same technique manually validated earlier via curl for
 * hand-built sample routes, now called directly from the app.
 */
object BRouterClient {
    private const val BASE_URL = "https://brouter.de/brouter"
    private const val PROFILE = "trekking"

    /** Returns the raw routed GPX text following roads through [waypoints], in order. */
    suspend fun route(waypoints: List<RouteWaypoint>): Result<String> = withContext(Dispatchers.IO) {
        if (waypoints.size < 2) {
            return@withContext Result.failure(IllegalArgumentException("Need at least 2 waypoints to route"))
        }
        try {
            val lonlats = waypoints.joinToString("|") { "${it.lon},${it.lat}" }
            val query = "lonlats=${URLEncoder.encode(lonlats, "UTF-8")}" +
                "&profile=$PROFILE&alternativeidx=0&format=gpx"
            val url = URL("$BASE_URL?$query")
            val connection = url.openConnection() as HttpURLConnection
            connection.requestMethod = "GET"
            connection.connectTimeout = 15_000
            connection.readTimeout = 30_000
            try {
                val responseCode = connection.responseCode
                if (responseCode != HttpURLConnection.HTTP_OK) {
                    val errorBody = connection.errorStream?.bufferedReader()?.use { it.readText() }
                    return@withContext Result.failure(
                        IOException("BRouter request failed ($responseCode): ${errorBody ?: "no details"}"),
                    )
                }
                val body = connection.inputStream.bufferedReader().use { it.readText() }
                Result.success(body)
            } finally {
                connection.disconnect()
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}
